File: ruby/lib/token.rb
# frozen_string_literal: true

module Token
  Illegal = Data.define
  Eof = Data.define

  Ident = Data.define(:ident)
  Int = Data.define(:int)
  Assign = Data.define
  Plus = Data.define
  Minus = Data.define
  Bang = Data.define
  Asterisk = Data.define
  Slash = Data.define

  NotEq = Data.define
  Equal = Data.define

  Lt = Data.define
  Gt = Data.define

  Comma = Data.define
  Semicolon = Data.define
  LParen = Data.define
  RParen = Data.define
  LSquirly = Data.define
  RSquirly = Data.define

  Function = Data.define
  Let = Data.define
  True = Data.define
  False = Data.define
  If = Data.define
  Else = Data.define
  Return = Data.define
end

File: haskell/src/Token.hs
module Token where

import Data.Char (isLetter)

type Tokenizer = String -> [Token] -- type synonym. A function from String to Linked list of Token

data Token           -- enum Token {
    = Illegal        --  Illegal,
    | Eof            --  Eof,
    | Ident String   --  Ident(String),
    | Int String     -- ...
    | Assign
    | Plus
    | Minus
    | Bang
    | Asterisk
    | Slash
    | LessThan
    | GreaterThan
    | Equal
    | NotEqual
    | Comma
    | Semicolon
    | LParen
    | RParen
    | LSquirly
    | RSquirly
    | Function
    | Let
    | TrueTok
    | FalseTok
    | If
    | Else
    | Return
    deriving (Show, Eq, Ord)  -- #[derive(Debug, PartialEq, Eq)]

isIdentChar :: Char -> Bool
isIdentChar c = isLetter c || c == '_' -- valid identifier characters

identToken :: String -> Token    -- Simple Pattern matching
identToken "fn" = Function
identToken "let" = Let
identToken "true" = TrueTok
identToken "false" = FalseTok
identToken "if" = If
identToken "else" = Else
identToken "return" = Return
identToken x = Ident x
